### PR TITLE
withinCompare handles Relative and negative values

### PR DESCRIPTION
**Pull request by jmedding**
In response to issue #211(Expect.within for Relative does not work for negative values).
- no changes needed to Absolute tolerance handling.
- solution was to adapt Relative tolerance handling to work like Absolute
-- calculate actual tolerance and then subtract it on the low end and add it to the high end

### CODE REVIEW

File: elm-package.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-    "version": "4.2.0",
+    "version": "4.2.1",
```
Comment (mgold) [inline elm-package.json]: Don't bump  the project version in a PR. Leave that for the maintainers. (Although we will almost certainly release a patch fix with this.)

File: tests/FloatWithinTests.elm [hunk 2]
```diff
@@ -35,6 +35,16 @@ floatWithinTests =
...
                             |> Expect.within (Relative 0.0001) (radius * 3.14)
             ]
+        , describe "negative use-cases"
+            [   test "negative floats with Absolute" <|
+                    \_ -> -2.9 |> Expect.within (Absolute 0.1) -3
+            ,   test "negative floats with Relative" <|
+                    \_ -> -2.9 |> Expect.within (Relative 0.1) -3
+            ,   test "negative floats with AbsoluteOrRelative" <|
+                    \_ -> -2.9 |> Expect.within (AbsoluteOrRelative 0.1 0.1) -3                
+            ,   test "negative floats with AbsoluteOrRelative passint with Relative" <|
```
Comment (mgold) [inline tests/FloatWithinTests.elm]: Typo, 'passint' (passing?).

**Comment by jmedding** [comment]
Thanks for the comments.  I  have made the changes you proposed.

File: tests/FloatWithinTests.elm [hunk 3]
```diff
@@ -35,6 +35,16 @@ floatWithinTests =
                         (radius * pi)
                             |> Expect.within (Relative 0.0001) (radius * 3.14)
             ]
+        , describe "negative use-cases"
```
Comment (drathier) [inline tests/FloatWithinTests.elm]: Is the use-case negative? How about `negative arguments`?

File: tests/FloatWithinTests.elm [hunk 4]
```diff
@@ -35,6 +35,16 @@ floatWithinTests =
...
+            [   test "negative floats with Absolute" <|
+                    \_ -> -2.9 |> Expect.within (Absolute 0.1) -3
+            ,   test "negative floats with Relative" <|
+                    \_ -> -2.9 |> Expect.within (Relative 0.1) -3
+            ,   test "negative floats with AbsoluteOrRelative and pass on Absolute" <|
+                    \_ -> -2.9 |> Expect.within (AbsoluteOrRelative 0.1 0.0001) -3                
+            ,   test "negative floats with AbsoluteOrRelative and pass on Relative" <|
+                    \_ -> -2.9 |> Expect.within (AbsoluteOrRelative 0.001 0.05) -3   
+            ]
```
Comment (drathier) [inline tests/FloatWithinTests.elm]: I'd like some test-cases where just one of the arguments is negative.

File: src/Expect.elm [hunk 5]
```diff
@@ -813,7 +813,7 @@ withinCompare tolerance a b =
             a - absolute tolerance <= b && b <= a + absolute tolerance
 
         withinRelativeTolerance =
-            (a * (1 - relative tolerance) <= b && b <= a * (1 + relative tolerance))
-                || (b * (1 - relative tolerance) <= a && a <= b * (1 + relative tolerance))
+            (a  - (abs (a * relative tolerance)) <= b && b <= a + (abs (a * relative tolerance)))
+                || (b - (abs (b * relative tolerance)) <= a && a <= b + (abs (b * relative tolerance)))
     in
-    (a == b) || withinAbsoluteTolerance || withinRelativeTolerance
+        (a == b) || withinAbsoluteTolerance || withinRelativeTolerance
```
Comment (drathier) [inline src/Expect.elm]: I'm not entirely convinced that just adding `abs` does the correct thing here, e.g. when only one argument is negative. It looks fine, but I'd like some tests to be sure.

**Comment by drathier** [comment]
Thought. Do we have a similar bug for `Absolute` tolerances as well?

File: tests/FloatWithinTests.elm (hunk 3 above)
Comment (jmedding) [reply tests/FloatWithinTests.elm]: We don't want a negative tolerance, which is one of the arguments, so I went with "use-cases with negative nominal and/or actual values"

File: src/Expect.elm (hunk 5 above)
Comment (jmedding) [reply src/Expect.elm]: Think of a number line. Take your nominal value(n) and and the absolute value of the tolerance(t). Then check if the actual value is between n-t and n+t.  This is the way that absolute worked from the beginning and it is very simple and robust.  Relative was using (1-t)*n, which works for simple cases, but gets all weird when n is negative or t is greater than one because it moves the tolerance band in the opposite direction.  All this is to say that I didn't just add 'abs', but changed the algorithm to eliminate the source of the bug.  By the way, my initial solution was to use a ` case` statement to check if the nominal value was less then zero and then use a branch where all the comparisons where flipped around to work with the inversion caused by the negative values, but this caused the communatative tests to fail - fuzz was send all kinds of weird values, like 0 for nominal, or tolerance greater than 1, which triggered all the edge-case failure modes in the original approach.  This led me to my current solution.

Comment (jmedding) [reply src/Expect.elm]: I hope this answers your question below as well about having a similar issue in absolute.  I did tests for Absolute tolerances as well, and they all passed without changes.

**Comment by mgold** [comment]
My superficial concerns have been addressed, so I'm leaving this in @drathier's capable hands.

File: tests/FloatWithinTests.elm [hunk 6]
```diff
@@ -35,15 +35,29 @@ floatWithinTests =
...
+            ,   test "negative nominal and positive actual with Absolute" <|
+                    \_ -> 0.001 |> Expect.within (Absolute 3.3) -3
+            ,   test "negative nominal and positive actual with Relative" <|
+                    \_ -> 0.001 |> Expect.within (Relative 1.1) -3
+            ,   test "negative actual and positive nominal with Absolute" <|
+                    \_ -> -0.001 |> Expect.within (Absolute 3.3) 3
+            ,   test "negative actual and positive nominal with Relative" <|
+                    \_ -> -0.001 |> Expect.within (Relative 1.1) 3
+            ,   expectToFail <|
+                    test "negative nominal should fail as avtual is close, but positive with Absolute" <|
```
Comment (drathier) [inline tests/FloatWithinTests.elm]: Minor typos: do a search for "avtual" :)

Comment (jmedding) [reply tests/FloatWithinTests.elm]: Nice catch - fixed.

**Comment by drathier** [comment]
Awesome!
